extensions/neo4j/runtime/src/main/java/io/quarkus/neo4j/runtime/JTAAwareDriver.java

```diff
@@ -0,0 +1,166 @@
+package io.quarkus.neo4j.runtime;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.transaction.Status;
+import javax.transaction.Synchronization;
+import javax.transaction.TransactionSynchronizationRegistry;
+
+import org.neo4j.driver.AccessMode;
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.Metrics;
+import org.neo4j.driver.Session;
+import org.neo4j.driver.SessionConfig;
+import org.neo4j.driver.async.AsyncSession;
+import org.neo4j.driver.reactive.RxSession;
+import org.neo4j.driver.types.TypeSystem;
+import org.neo4j.driver.util.Experimental;
+
+import io.quarkus.arc.Arc;
+
+final class JTAAwareDriver implements Driver {
+
+    private static final Object SESSION_KEY = new Object();
+    private final Driver delegate;
+
+    JTAAwareDriver(Driver delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean isEncrypted() {
+        return delegate.isEncrypted();
+    }
+
+    @Override
+    public Session session() {
+
+        return this.session(SessionConfig.defaultConfig());
+    }
+
+    @Override
+    public Session session(SessionConfig sessionConfig) {
+
+        var synchronizationRegistryHandle = Arc.container().instance(TransactionSynchronizationRegistry.class);
+        if (synchronizationRegistryHandle.isAvailable()) {
+
+            var synchronizationRegistry = synchronizationRegistryHandle.get();
+            if (synchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE) {
+
+                var session = (Session) synchronizationRegistry.getResource(SESSION_KEY);
+                if (session != null) {
+                    return session;
+                }
+
+                return createAndRegisterJTAAwareSession(synchronizationRegistry, sessionConfig);
+            }
+        }
+
+        return this.delegate.session(sessionConfig);
+    }
+
+    Session createAndRegisterJTAAwareSession(TransactionSynchronizationRegistry synchronizationRegistry,
+            SessionConfig sessionConfig) {
+
+        var configBuilder = SessionConfig.builder();
+        // We will use write mode by default here… We have no way to determine this from the transactional annotation
+        // If this is unwanted, make sure you use @Transactional(NEVER) and create a config on you own.
+        configBuilder = configBuilder.withDefaultAccessMode(AccessMode.WRITE)
+                .withBookmarks(sessionConfig.bookmarks());
+        configBuilder = sessionConfig.database().map(configBuilder::withDatabase).orElse(configBuilder);
+        configBuilder = sessionConfig.fetchSize().map(configBuilder::withFetchSize).orElse(configBuilder);
+
+        var session = new JTAAwareSession(this.delegate.session(configBuilder.build()));
+
+        synchronizationRegistry.putResource(SESSION_KEY, session);
+        synchronizationRegistry.registerInterposedSynchronization(new Synchronization() {
```

Comment:
This is not my area of expertise, but I don't really understand the intended semantics here.

You really mean to commit to Neo4J in an _afterCompletion_ callback of an interposed synchronization?

Is that how Neo4J integrates with other systems as well, or do you have plans to improve it later?

I'm not strictly against merging it - your call :) But I wonder if it warrants a note about safety in the documentation.

@ochaloup could you have a look as well please?

Reply:
I am quite new to the JTA semantics here and I tried to understand what the Mongo extension does. 

What would be the alternative? In Spring world I am used to run my own transaction manager.

Reply:
> The afterCompletion callback will be invoked in an undefined 
     * context. No access is permitted to "user components"
     * as defined above. Resources can be closed but no transactional
     * work can be performed with them.

From the Java Doc about interposed synchronization. Yeah, I think you're right…

Reply:
But: `beforeCompletion` seems only to receive status `active`. I'll keep it as is until someone with more expertise tells me more :)

Reply:
>  In Spring world I am used to run my own transaction manager.

What do you mean by "my own" ?

Reply:
@mmusgrov , @ochaloup, @tomjenkinson could someone from Narayana expertise provide guidance please?

Reply:
You would need to get hold of the transaction manager (`com.arjuna.ats.jta.TransactionManager.transactionManager()`) and use that to get the currently active transaction (`javax.transaction.Transaction`) which has methods to enlist resources. Neo4j would need to pass in an XAResource to that method and we would use that to notify it when to prepare/commit/rollback). Neo4j would also need to give us something to manage recovery with.

We (the narayana project) did a JTA integration with the Neo4j team many years ago but I am informed that that support was removed (not sure though).

Reply:
Thanks a lot @mmusgrov I think I can work with that.

I work in the Neo4j drivers team, not in the (embeddable) product itself, so I can only use the driver here. 

I wonder if the same concerns would apply to the Mongo extension? I can understand those concerns and I understand that the synchronization is not optimal todo this.

For the time being i convert this to a draft then.

Reply:
In April last year their support portal said: "MongoDB supports multi-document ACID transactions within a single replica set (4.0+) or sharded cluster (4.2+). XA transactions are not a current feature".

In other words MongoDB can support updates to multiple MongoDB documents but that is as far as it goes.

But I will take a look at that Mongo extension to see what it is doing.

Reply:
> I wonder if the same concerns would apply to the Mongo extension? 

Right, thanks for pointing it out - we'll need to review that as well. However with MongoDB I don't think people can expect much in terms of transactionality.

Reply:
@michael-simons Will you provide me with a link to the `Mongo extension` that you mentioned, I can only find `mongodb-client` which does not appear to do much with transactions?

Reply:
Sure @mmusgrov : https://github.com/quarkusio/quarkus/blob/2.2.3.Final/extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java#L330-L383

Reply:
@loicmathieu Hi Loic, apologies for pulling you in here but Michael is applying a similar approach to session/transaction handling as was done in the [Mongo extension](https://github.com/quarkusio/quarkus/blob/2.2.3.Final/extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java#L330-L383). We would like to understand what the synchronization code in MongoOperations is trying to achieve: the code is registering an (interposed) synchronisation which aborts or commits a session in the `afterCompletion` callback. We picked up on this in the review of the Neo4j PR since trying to terminate a session from an afterCompletion callback that is not guaranteed to be called. Could you help clarify what's going on with this code and if possible could you also comment on what Michael is doing in his changes to the Neo4j driver and whether it looks sensible (on my first reading of the PR it looked like the goal was to join with the transaction but my presumption is probably wrong so any help that you can provide will be appreciated).

Reply:
The goal is to participate with a Neo4j transaction in the cycle of a managed JTA transaction. I want to begin a transaction when the managed one begins and close it accordingly. 

The dance with the driver Interfaces disables our own transaction manager and basically makes our auto commit transaction a managed one. 

Judging from what I learned here writing a proper resource would be correct. Whether I can support recovery from within our client driver is another question.

Reply:
Hi,

The design of MongoDB transactions has been discussed in multiple places and there was several attemps to implement them. The current impl is based on @FroMage advices and I think also some guidelines from @Sanne. The principle is to contribute to the global JTA transaction via the synchronization mechanism. 
If we are in a transaction context (so if a transaction is alive) we get the current Session or create a new one and register it inside the tx manager, see https://github.com/quarkusio/quarkus/pull/14002/files#diff-d003b221f9b819da503954b1f03c63b8ddc7fb18b78c6921cc9ae4d9e1483d1aR334
When the transactions ends, we commit or rollback based on its state, see https://github.com/quarkusio/quarkus/pull/14002/files#diff-d003b221f9b819da503954b1f03c63b8ddc7fb18b78c6921cc9ae4d9e1483d1aR355

>  trying to terminate a session from an afterCompletion callback that is not guaranteed to be called.

I was not aware of this, I'm not a JTA expert and this was the guideline I was given to use. If it's the case the MongoDB implementation is wrong. I do some testing (and we have integration test) but not very extensive. I didn't have any feedback from any user that it didn't work either ...

We do enlist MongoDB ressource to the txmanager, we enlist the MongoDB session (this is the transaction object) so we participate in the global tx in a somewhat standard way.

To understand more where we are comming for MongoDB you can have a look at the PR https://github.com/quarkusio/quarkus/pull/14002 and go back to the two previous PR with previous failed attempt to implement it.

Reply:
Thanks for the links, @loicmathieu 

Can you please point me to where you enlist "enlist MongoDB ressource to the txmanager"? Like in enlisting a XAResource or just adding it to the synchronization?

Reply:
We just add the MongoDB session as a resource inside the txmanager then register and interposed synchronization: https://github.com/quarkusio/quarkus/pull/14002/files#diff-d003b221f9b819da503954b1f03c63b8ddc7fb18b78c6921cc9ae4d9e1483d1aR348 this is what I called "enlist MongoDB ressource to the txmanager".
MongoDB is not XA compatible.